### DISCUSSION TITLE
How to stay under partition limit

### DISCUSSION THREAD

**Discussion by khanakia** [ClickHouse/ClickHouse#73630] [Q&A]
As click house suggests 100000 parts across all databases for a service here https://clickhouse.com/docs/knowledgebase/maximum_number_of_tables_and_databases

I can understand and somehow control the number or partition but how do I control the parts?

I have a `tasks` and `task_records` and I have partitioned `PARTITION BY task_id` 

I have around only 1 million records and it seems to have 300 parts already for one `task_id`. It seems I will easily run out of `100000 parts`

So what's the use of Part at all and how do I stay under I can't understand.

![Screenshot 2024-12-19 at 8 23 40 PM](https://github.com/user-attachments/assets/af122d44-6c86-4f94-aeb2-c2c525863943)

**Comment by inv2004** [comment]
@khanakia sometimes a lot of parts means a lot of inserts. Probably you should check if you use batch insert, or, if it not possible in service side - to use some CH methods like async inserts or Buffer engine

**Reply by khanakia** [reply]
Yes i use batch inserts.

So should i avoid using batch inserts and just use Async inserts ?

**Comment by inv2004** [comment]
Probably just check how many inserted you do

**Reply by khanakia** [reply]
the problem is we sync data from Google Ads. However, we need to fetch the Google Ads data again for the last 3 months and because Google Ads keeps refreshing their past data up to 3 months.

But Updating or Deleting the data is expensive in clickhouse, so we thought to partition data by Date, delete the partitions for 3 months, and resync the data.

But as Clickhouse has a maximum number of partition limits and we can have thousands of Customers so this way we can easily run of partition limit.

So what do you suggest how should we structure the database?

**Comment by pengxiaoo** [comment]
so you encountered performance issue in clickhouse. in fact the num of parts could not be the root cause(it is only an consequences), neither your insertion manner.

the root cause must be your primary key and partition key. can you illustrate your table DDL so we can check?

you mentioned you have 1 million records and you partitioned the records by task_id. how many unique task_ids do you have? high cardinality field is not suitable to be used as partition key

for the google ads data sync, partition data by date is not a good idea, the granularity is too fine which will bring trouble. unless you have super large volume for a single day, you should partition by month
